packages/gatsby: redux/nodes.ts: Fix binarySearch used in non-sift lte implementation. Add tests.

reltuk opened the pull request:
<!--
  Have any questions? Check out the contributing docs at https://gatsby.dev/contribute, or
  ask in this Pull Request and a Gatsby maintainer will be happy to help :)
-->

<!--
  Is this a blog post? Check out the docs at https://www.gatsbyjs.org/contributing/blog-contributions/, and please mention if the blog post is pre-approved
  by someone from Gatsby.
-->

## Description

When we upgraded gatsby recently our build broke with a hanging node process pegging a single CPU inside our `createPages` step. We tracked it down to a query that used an `lte` filter and the `binarySearch` function that was introduced here: https://github.com/gatsbyjs/gatsby/commit/fd57224888. The pivot there is set to `Math.floor((max - min) / 2)`, which is probably meant to be `min + Math.floor((max - min) / 2)`.

This is a small fix to prevent the infinite loop.

Includes some tests of run-sift that triggered the infinite loop in the old path and no longer do.

### Documentation

N/A

## Related Issues

Not aware of an issue filed in regards to this.

reltuk commented on packages/gatsby/src/redux/nodes.ts:
```diff
@@ -419,37 +419,25 @@ function addNodeToBucketWithElemMatch(
-
-  // Shouldn't be reachable, but just in case, fall back to Sift if so.
-  return undefined
+  return -(min + 1)
```
This only had one caller and I kind of struggled to understand the semantics of the previous method --- sometimes it would return `[i, i+1]` when the element was present in the array and sometimes it would return `[i, i]`. I changed it to just return a value that definitively identifies the result of the search.

pvdz replied:
So my thought here was that this binary search would be more likely to miss than to hit. In fact, it's a very outlier edge case (caused by coercion) if it would hit, so I would want to receive the two indexes between where the pivot ended.

If this, for whatever reason, just outright failed, the fallback is to go through Sift anyways. Sift is the original implementation that we're slowly replacing / making faster. The caller is checking explicitly for `undefined` and will bail entirely on the fast indexes if this search returns it.

That said, it's not pretty and I don't think it should fail so I'm not against improving that situation.

Admittedly I'm not super fond of returning a negative number in case of a miss but I'm also aware that any alternative is just slower, including returning an array like it does now. Can't win here :p

reltuk commented on packages/gatsby/src/redux/nodes.ts:
```diff
@@ -512,29 +500,20 @@ export const getNodesFromCacheByValue = (
-    // So we have to consider weak comparison and may have to include the pivot
-    const until = pivotValue <= filterValue ? inclPivot : exclPivot
-    return new Set(nodes!.slice(0, until))
+    if (idx >= 0) {
```
I was confused by the further tests on the returned indexes from `binarySearch` and the indirection through `ranges!.get(pivotValue)`. This changed code just returns the set based on the range found based on the binary search. This maybe misses some cache interaction or something? Happy for guidance here...

pvdz replied:
The reason for this indirection is that this approach is to be reused for `<` and `<=`. One needs to slice up to the pivot, the other will need to also include all the nodes for given value (the values array is most likely not unique).

I think that this is necessary for both cases and I'm surprised this would work for you.

For `lt` you need it for the case like `[1,2,2,2,2,3]`, where binary search would just return the middle index, which is not what you should slice on because you want to slice `[1]`, not `[1,2,2]`.
For `lte` you would want to receive `[1,2,2,2,2]`, not  `[1,2,2]`.

The indirection takes the value, looks up the precomputed range for all nodes with that value, and then slice based on that range, and whether the pivot was found (include the range) or not (exclude the range).

I hope this clarifies it a bit. If that makes sense, can you help me to confirm whether or not your approach should work? Because I feel it shouldn't, but I expect that you ran the tests and it did work.

(Looking at it now, and considering a 1:1 index match is kinda useless, the set of ordered values could be a _set_ (of uniques) which would speed up the binary search in some cases)

reltuk commented on packages/gatsby/src/redux/__tests__/run-sift.js:
```diff
@@ -131,6 +131,17 @@ if (!process.env.GATSBY_DB_NODES || process.env.GATSBY_DB_NODES === `redux`) {
         actions.createNode(node, { name: `test` })(store.dispatch)
       )
     })
+    // There was once a bug in filter lte evaluation that only triggered under
```
I couldn't trigger the failing case with just the 4 nodes in the original test. Added this machinery to add a few more, which did get me failing tests before the changes. Happy for guidance here.

pvdz replied:
I've added a similar approach in #23637 to the run-query tests. I guess, since the number of nodes were this small, the pivot never really had to be set again and it didn't lead to a problem. The set was just too small. So I've added some range checks to a set of 100 nodes as well (and checked that they caught this infinite loop as well before fixing it). Thanks for the inspiration.

I feel we don't have to add it here anymore. What do you think?

pvdz commented:
Dang, looking into this. Thank you!

pvdz commented on packages/gatsby/src/redux/nodes.ts:
```diff
@@ -419,37 +419,25 @@ function addNodeToBucketWithElemMatch(
   let max = values.length - 1
-  let pivot = Math.floor(values.length / 2)
   while (min <= max) {
+    const pivot = min + Math.floor((max - min)/ 2)
```
:+1:

pvdz commented on packages/gatsby/src/redux/nodes.ts:
```diff
@@ -512,29 +500,20 @@ export const getNodesFromCacheByValue = (
+    if (idx >= 0) {
+	// We unexpectedly found the value in the array. Inlucude up through
+	// its index in the returned range.
+        return new Set(nodes!.slice(0, idx+1))
```
I think I need to add a test for `[1,2,2,2,2,3]` which should fail with this appraoch. Right?

reltuk replied:
The previous approach was definitely broken for this case. I think the suggested test would be a good one. But I think attempts to test this path might be hard at the level of filterNodes because the equality filter will probably fast-path the evaluation before we hit the binarySerach?

pvdz replied:
If we can't reach it it shouldn't be in there ;) You can use coercion, that's the reason we have to support direct hits here in the first place.

So a node `{x: 5}` and a filter `{x:{ lte: '5'}}`, or the num/str flipped, should do the trick here.

I tend to add these filter tests in `__tests__/run-sift.js` as they feel more concise there, but that's up to you.

reltuk replied:
I see...`filterCache.meta.values{De,A}sc`  are already unique values, so we always need to do the `ranges!` indirection. My brokenness wasn't caused by any existing tests, but I added some non-type-coercion tests that do catch it.

When I went to write the specific test with type coercion, I got different behavior between the non-sift and the sift path. In particular, sift on `[{x: 1}, {x: 2}, {x: 2}, {x: 3}]` filter `{x:{ lte: '2'}}` seems to return no nodes at all. I wrote added the tests with assertion for the two different cases. Maybe not what we want...

pvdz commented:
Ok, hotfix published in gatsby@2.21.7, thank you for bringing it to our attention.

I've gone through this PR thoroughly. I agree that the search can be improved but I'm not sure the PR covers it. I've tried to be thorough in my explanation and hope they make sense. Please let me know if something is unclear and let's try to improve the search where possible.

reltuk commented on packages/gatsby/src/redux/nodes.ts:
```diff
@@ -419,37 +419,25 @@ function addNodeToBucketWithElemMatch(
-      return [min, max]
-    }
-
-    pivot = Math.floor((max - min) / 2)
```
Totally the right thing to do! Thank you for getting the fix out so fast :).

reltuk commented on packages/gatsby/src/redux/nodes.ts:
```diff
@@ -419,37 +419,25 @@ function addNodeToBucketWithElemMatch(
       // Move pivot to middle of nodes left of current pivot
       // assert pivot < max
-      max = pivot
+      max = pivot - 1
```
I've restructured things a bit after your feedback, but this feature remains.

The basic answer is: yes. If it does, then the while loop will exit on the next iteration, since it necessarily means `min > max` (`min` is always `>= 0` and max is always `<= values.length - 1`). This is totally safe. If `max==-1` at the end of the loop, then `pivot==0` at the top of the loop, which means `min==0` and `max==1 || max==0`. In either case, `min` remains `0` and the index we're looking for is `0`. In the other case, if `min > values.length - 1` at the end of the loop, then `pivot == values.length - 1` at the top of loop and the predicate is still `true`, so the index we're looking for is actually `values.length`.

reltuk commented:
@pvdz Thank you for such a great review! Really helped me better understand what factors were at play in the existing implementation and how my initial approach was missing some details.

I think I found a way to clean things up a bit and even "win" with the case of the negative return as well. I changed `binarySearch` so it never fast-path returns in the case of `needle == values[pivot]`. Instead it searches all the way to the end condition, where `min > max`, which necessarily means `min` is the index of the value in the array where the predicate switches from `true` to `false`. I think that's exactly the index that the call site wants to `slice` until.

Also removed the tests, which I do think are unnecessary at this point.

reltuk commented on packages/gatsby/src/redux/nodes.ts:
```diff
@@ -452,75 +452,32 @@ function addNodeToBucketWithElemMatch(
+// portion of `values`. If all `values` match, returns
+// `values.length`. If no `values` match, returns `0`, the first index
+// where `needle` is false.
+function binarySearch<T>(values: Array<T>, pred: (cur: T) => boolean): number {
```
It's possible we want to leave these specialized for `lte` and `gte` instead of using this predicate approach. Happy to change it.

reltuk commented:
Force pushed a rebase because my branch got a bit messy. Hope that's OK.

reltuk commented:
So, I generally like the approach laid out here, but I admit that I don't totally understand the preconditions on the `binarySearch` call, so it's possible that this doesn't cover all bases.

In particular, it's not clear to me if the input array is always actually well behaved on the incoming predicate. Well behaved here means: there is a possibly empty prefix of the array where the predicate evaluates to `true`, and then for every element in the array after that prefix the predicate evaluates to `false`. If it's possible that isn't true because of type coercion or `NaN`s or something, then this approach is not going to always generate correct results. And it doesn't have any mechanism or anything to fall back to sift.

If the above condition isn't true, I'm not certain the previous approach always fell back appropriately either.

The `binarySearch` will always find a place in the array where `pred` transitions from `true` to `false`, it just might not be a unique place, and it certainly implies we're not getting the semantics we want.

reltuk commented:
An example of a non-well-behaved input array is:

`[' ', '-1', 'a', 'z', 'A', 'Z', '1', '1z', '2']`,

which sorts descending (per `entries.sort(([a], [b]) => (a > b ? -1 : a < b ? 1 : 0))`, from `redux/nodes.ts`) as

`["z", "a", "Z", "A", "2", "1z", "1", "-1", " "]`.

Given the predicate `gte: 0` it will evaluate, at each position in the sorted values array:

`[false, false, false, false, true, false, true, false, true]`.

```javascript
[' ', '-1', 'a', 'z', 'A', 'Z', '1', '1z', '2'].sort((a, b) => (a > b ? -1 : a < b ? 1 : 0)).map(function(n) { return n >= 0; })
> (9) [false, false, false, false, true, false, true, false, true]
```

reltuk commented:
Gonna close this. Feel free to reach out if you're still interested in it.

Thank you for the quick turn around on the fix!
